service/history/workflowRebuilder.go

```diff
@@ -0,0 +1,210 @@
+		newStateRebuilder nDCStateRebuilderProvider
+		transaction       workflow.Transaction
+		logger            log.Logger
+	}
+)
+
+var _ workflowRebuilder = (*workflowRebuilderImpl)(nil)
+
+func NewWorkflowRebuilder(
+	shard shard.Context,
+	workflowCache workflow.Cache,
+	logger log.Logger,
+) *workflowRebuilderImpl {
+	return &workflowRebuilderImpl{
+		shard:         shard,
+		workflowCache: workflowCache,
+		executionMgr:  shard.GetExecutionManager(),
+		newStateRebuilder: func() nDCStateRebuilder {
+			return newNDCStateRebuilder(shard, logger)
+		},
+		transaction: workflow.NewTransaction(shard),
+		logger:      logger,
+	}
+}
+
+func (r *workflowRebuilderImpl) rebuild(
+	ctx context.Context,
+	workflowKey definition.WorkflowKey,
+) (retError error) {
+	context, releaseFn, err := r.workflowCache.GetOrCreateWorkflowExecution(
+		ctx,
+		namespace.ID(workflowKey.NamespaceID),
+		commonpb.WorkflowExecution{
+			WorkflowId: workflowKey.WorkflowID,
+			RunId:      workflowKey.RunID,
+		},
+		workflow.CallerTypeAPI,
+	)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		releaseFn(retError)
+		context.Clear()
+	}()
+
+	msRecord, dbRecordVersion, err := r.getMutableState(workflowKey)
+	if err != nil {
+		return err
+	}
+	requestID := msRecord.ExecutionState.CreateRequestId
+	versionHistories := msRecord.ExecutionInfo.VersionHistories
+	currentVersionHistory, err := versionhistory.GetCurrentVersionHistory(versionHistories)
+	if err != nil {
+		return err
+	}
+	branchToken := currentVersionHistory.BranchToken
+	stateTransitionCount := msRecord.ExecutionInfo.StateTransitionCount
+
+	rebuildMutableState, rebuildHistorySize, err := r.replayResetWorkflow(
+		ctx,
+		workflowKey,
+		branchToken,
+		stateTransitionCount,
+		dbRecordVersion,
+		requestID,
+	)
+	if err != nil {
+		return err
+	}
+	return r.persistToDB(rebuildMutableState, rebuildHistorySize)
+}
+
+func (r *workflowRebuilderImpl) replayResetWorkflow(
+	ctx context.Context,
+	workflowKey definition.WorkflowKey,
+	branchToken []byte,
+	stateTransitionCount int64,
+	dbRecordVersion int64,
+	requestID string,
+) (workflow.MutableState, int64, error) {
+
+	rebuildMutableState, rebuildHistorySize, err := r.newStateRebuilder().rebuild(
+		ctx,
+		r.shard.GetTimeSource().Now(),
+		workflowKey,
+		branchToken,
+		math.MaxInt64-1, // NOTE: this is last event ID, layer below will +1 to calculate the next event ID
+		nil,             // skip event ID & version check
+		workflowKey,
+		branchToken,
+		requestID,
+	)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	// note: this is an admin API, for operator to recover a corrupted mutable state, so state transition count
+	// should remain the same, the -= 1 exists here since later CloseTransactionAsSnapshot will += 1 to state transition count
+	rebuildMutableState.GetExecutionInfo().StateTransitionCount = stateTransitionCount - 1
+	rebuildMutableState.SetUpdateCondition(rebuildMutableState.GetNextEventID(), dbRecordVersion)
```

Comment:
If next eventID in db is corrupted, does it mean the update will fail? 
Shall we use the nextEvenID in `msRecord` here?

Reply:
we stop using next event ID as CAS condition ~6 months ago, we should do more cleanup.